Compiler crash: declare-then-assign-in-try local crashes SymbolSerializer

## Summary

The idiom

```csharp
Dictionary x;
try { x = JSON.Parse(s); }
catch { return; }
// use x
```

works in `RpcConnection.OnWsMessage` (`src/Client/Services/ApiServices/Rpc/RpcConnection.cs:163-168`) but the *exact same pattern* in `WebSocketLogSink.SendPayload` (same project, same usings, same target framework) crashes Cs2Jsc with `NotSupportedException` at `SymbolSerializer.Serialize(TypeSymbol)`. The crash is binding-order sensitive.

**Priority:** Medium — rarer than #93 but more bewildering when it hits because identical-looking code works in a sibling file.

**Related:** #93 (Bug #1 — same `SymbolSerializer.Serialize` crash site, different trigger pattern)

## Environment

| | |
|---|---|
| Cs2Jsc tool | 1.1.3 (host) |
| NScript SDK | 1.1.1 |
| Sunlight.Framework + SunlightUnit + SunlightTestAdapter | 1.1.1 |
| mscorlib (Mcqdb.NScript.MsCorlib) | 1.1.1 |
| Target | netstandard2.1, C# 8.0 |
| .NET SDK | 9.0.308 |
| Repro repo | AgentWorkBench at `B:\sources\AgentWorkBench\` |

## Steps to Reproduce

### File A — compiles successfully

```csharp
namespace Repro {
    using System.Web;
    public sealed class A {
        public void Do(string text) {
            Dictionary dict;
            try { dict = JSON.Parse(text); }
            catch { return; }
            // ... use dict ...
        }
    }
}
```

### File B — crashes Cs2Jsc 1.1.1

```csharp
namespace Repro {
    using System.Web;
    public sealed class B {
        private readonly object _whatever;
        public B(object x) { _whatever = x; }

        public void Send(string payload) {
            Dictionary argsObj;
            try { argsObj = JSON.Parse(payload); }
            catch { return; }
            DoSomething(argsObj);
        }

        private static void DoSomething(object o) { }
    }
}
```

In the AgentWorkBench repro, file A is `RpcConnection.cs`, file B is `WebSocketLogSink.cs`. The two files differ in unrelated ways (number of fields, presence of a constructor with parameters, etc.). The specific trigger has not been isolated yet, but it appears to be a binding-order interaction with NScript's per-class type-spec table.

## Expected Behavior

Both files should compile identically — they use the same standard library API in the same way.

## Actual Behavior

```
Process terminated. System.NotSupportedException: Specified method is not supported.
```

### Stack trace (abridged) — identical to #93

```
at NScript.Csc.Lib.SymbolSerializer.Serialize(TypeSymbol type)
at NScript.Csc.Lib.BoundAstToAstBase.GetLocalVariable(LocalSymbol localSymbol, SerializationContext arg)
at NScript.Csc.Lib.BoundAstToAstBase.VisitLocalDeclaration(BoundLocalDeclaration node, SerializationContext arg)
…
```

## Workaround (used in production)

Inline the `JSON.Parse` and the consumer call inside a single `try`, so no `T x;`-then-assign-in-try local declaration is needed:

```csharp
private void SendPayload(string payload)
{
    try
    {
        var argsObj = JSON.Parse(payload);
        this._rpc.Notify("Diagnostics.LogBatch", argsObj);
    }
    catch { /* fire-and-forget */ }
}
```

**Trade-off:** if `Notify` itself throws (it shouldn't, but bugs happen), the catch will silently swallow that too. Acceptable for a fire-and-forget log sink, not in general.

## Suggested fix scope

In `NScript.Csc.Lib.SymbolSerializer.Serialize(TypeSymbol)`:

1. Enumerate the additional `TypeSymbol` subtypes that reach the default-throw arm — for this bug, likely a `LocalSymbol` whose `Type` is a synthesized union / unmaterialized inference produced by the try/catch binding path.
2. For each new subtype, either:
   - Serialize correctly (preferred), or
   - Promote the `NotSupportedException` into a Roslyn `Diagnostic` so the user sees a clean compile error pointing at the offending source line.

## References

- `src/Client/Services/ApiServices/Rpc/WebSocketLogSink.cs` (AgentWorkBench) — `SendPayload` uses the inlined-try idiom because of this bug; inline comment points here.
- `src/Client/Services/ApiServices/Rpc/RpcConnection.cs:163-168` — the same pattern that *does* compile, kept for reference comparison.

[Gautam's Dev bot]

<!-- BOT-PLAN v2 -->
## Revised design plan — WI-94: `SymbolSerializer.Serialize` crashes on `ErrorTypeSymbol` reaching method-signature serialization

**Revision driver:** reviewer asked for systematic debugging — regress the bug, identify the root cause, then propose a fix backed by evidence that the fix actually addresses the root cause. This plan replaces the speculative one in [#issuecomment-4513006537](https://github.com/achieveai/NScript/issues/94#issuecomment-4513006537).

---

### 1. Regression — minimal Roslyn-driven repro

Built a standalone harness (`/tmp/wi94-repro/`) that drives `NScript.Csc.Lib.SerializationHelper.InjectIntoCompilation` against the user's two patterns plus targeted variants. Temporary instrumentation enriched the `NotSupportedException` throw at `SymbolSerializer.cs:238` with `Kind` / `RuntimeType` / `FQN` / `ContainingType` of the offending `TypeSymbol`, plus a per-callback log of `methodSymbol` + `boundBody.HasErrors`.

Two scenarios reproduce a crash; one does not — the asymmetry pinpoints the root cause.

**Scenario B2 — reproduces the crash** (method `DoSomething(NotDefinedType o)`, signature contains an unresolved type, body is empty):

```
[WI-94 INSTR] OnBoundExpressionGenerated method='Repro.B2.Send(string)'           bodyHasErrors=True  initHasErrors=False
[WI-94 INSTR]   -> SKIPPED (HasErrors)
[WI-94 INSTR] OnBoundExpressionGenerated method='Repro.B2.DoSomething(NotDefinedType)' bodyHasErrors=False initHasErrors=False
CRASH: System.NotSupportedException
  Msg: SymbolSerializer.Serialize: unsupported TypeSymbol Kind='ErrorType',
       RuntimeType='Microsoft.CodeAnalysis.CSharp.Symbols.MissingMetadataTypeSymbol+TopLevel',
       FQN='object', ContainingType='<null>'.
  at NScript.Csc.Lib.SymbolSerializer.Serialize(TypeSymbol)            SymbolSerializer.cs:239
  at NScript.Csc.Lib.SymbolSerializer.AddTypeSpecId                    SymbolSerializer.cs:133
  at NScript.Csc.Lib.SymbolSerializer.GetTypeSpecSer                   SymbolSerializer.cs:126
  at NScript.Csc.Lib.SymbolSerializer.<Serialize>b__23_0(ParameterSymbol) SymbolSerializer.cs:324
  at NScript.Csc.Lib.SymbolSerializer.Serialize(MethodSymbol)          SymbolSerializer.cs:314
  at NScript.Csc.Lib.SymbolSerializer.GetMethodSpecId                  SymbolSerializer.cs:76
  at NScript.Csc.Lib.BoundAstToAstBase.GetMethodBody                   BoundAstToAstBase.cs:31
```

**Scenario B3 — does NOT crash** (local with unresolved type *inside* a method body — i.e. the literal "declare-then-assign-in-try" pattern in the issue title):

```
[WI-94 INSTR] OnBoundExpressionGenerated method='Repro.B3.Send(string)'  bodyHasErrors=True  initHasErrors=False
[WI-94 INSTR]   -> SKIPPED (HasErrors)
Emit success: False
Error count: 4    ← clean diagnostics, no crash
```

**Happy path — does NOT crash** (both file A and file B from the issue, with a complete-enough mscorlib facade): every method body has `HasErrors=False`, Serialize completes, Emit returns Success=False only because of unrelated facade nits. No `NotSupportedException` thrown.

### 2. Root cause

`OnBoundExpressionGenerated` is fired once per method by Roslyn's `MethodCompiler.CompileMethod`. The callback invokes `BoundAstToAstBase.GetMethodBody` which **first serialises the containing method's signature** via `GetMethodSpecId(methodSymbol)` (`BoundAstToAstBase.cs:31`) and **then** walks the body. Signature serialisation calls `SymbolSerializer.Serialize(MethodSymbol)` → `Serialize(TypeSymbol)` for the return type and each parameter type.

`SymbolSerializer.Serialize(TypeSymbol)` (`SymbolSerializer.cs:200-238`) has branches for `ArrayType` / `PointerType` / `TypeParameter` / `NamedType` and a fall-through `throw new NotSupportedException();`. It has **no `ErrorType` branch**.

When a method's signature references an unresolved type — distinct identifier in the user's project, a missing/mismatched assembly, or a metadata gap — Roslyn binds that type as `MissingMetadataTypeSymbol+TopLevel` (an `ExtendedErrorTypeSymbol` subclass with `Kind = ErrorType`). The body of *that* method may still be empty or otherwise well-formed, so `boundBody.HasErrors = false` and WI-93's `HasErrors` gate (already on `master` at 10458ded) does **not** fire. The signature walk then hits the fall-through arm and unwinds through Roslyn's body-compilation pipeline before `Compilation.Emit` can return its CS-coded diagnostics — the user sees a process crash with `NotSupportedException` instead of the underlying compile error.

The issue title's framing ("declare-then-assign-in-try local") is a **red herring**: B3 (literally that pattern with an unresolved local type) is *already* caught by WI-93's gate. The actual trigger in the user's repo is some other method in the same compilation whose signature carries an unresolved type — exactly the asymmetry the user described between `RpcConnection.cs` (compiles) and `WebSocketLogSink.cs` (crashes). It is not binding-order sensitive; it is signature-error sensitive.

### 3. Why WI-93's existing fix is insufficient

WI-93 (merged in PR #95) added two defences:
1. `boundBody.HasErrors` / `initializers.HasErrors` gate at the callback entry → skips bodies with bound errors.
2. Enriched `NotSupportedException` message at `SymbolSerializer.cs:238` → diagnostics on unsupported shapes.

Neither covers WI-94. The gate looks only at *body* errors; an empty body with an error-type *parameter* slips through (evidence: B2 trace above, `bodyHasErrors=False` followed by crash). The enriched throw is still a fatal unwind — useful for diagnosis but it doesn't let `Emit` complete.

### 4. Proposed fix — single change in `SymbolSerializer.Serialize(TypeSymbol)`

Add an `ErrorType` arm that returns a placeholder `TypeSpecSer` instead of throwing. The compilation already has CS-coded diagnostics for whatever produced the error type; Roslyn's `Emit` will return `Success=false` and the embedded `$$BstInfo$$` resource won't be consumed by stage 2 (the DLL won't be produced). The placeholder is never observed.

```csharp
// Sources/Compiler/NScript.Csc.Lib/SymbolSerializer.cs, inside Serialize(TypeSymbol)
// — placed BEFORE the existing `if (type.ContainingModule == null)` and
//   non-NamedType throw arms.
if (type.Kind == SymbolKind.ErrorType)
{
    // WI-94: a method signature or body referenced an unresolved type.
    // Roslyn has already produced CS-coded diagnostics; we just need to
    // avoid unwinding through SymbolSerializer with NotSupportedException
    // before Emit can return them. Return a sentinel; Emit will fail and
    // the AST resource will be discarded.
    return new TypeSpecSer
    {
        Name = "<error>",
        Namespace = null,
        Module = null,
    };
}
```

Keep the WI-93 `HasErrors` gate (defence in depth — avoids unnecessary body walks on already-broken methods). Keep the enriched-throw arm for *genuinely* unsupported shapes (function pointers, future Roslyn surface) so the next regression is identifiable from the message.

### 5. Evidence the proposed fix resolves the bug

Applied the patch above to `SymbolSerializer.cs` (commented out the temporary instrumentation in `SerializationHelper.cs` so the gate is identical to WI-93's behaviour on `master`) and re-ran the harness:

| Scenario | Before fix | After fix |
|---|---|---|
| File A + facade (happy) | Emit=False, 2 unrelated facade errors, **no crash** | Emit=False, 2 unrelated facade errors, **no crash** |
| File B + facade (happy) | Emit=False, 2 unrelated facade errors, **no crash** | Emit=False, 2 unrelated facade errors, **no crash** |
| File C workaround | Emit=False, 2 unrelated facade errors, **no crash** | Emit=False, 2 unrelated facade errors, **no crash** |
| **B2** (sig-level ErrorType) | **CRASH** `NotSupportedException` | **Emit=False, 4 CS errors, no crash** ✅ |
| B3 (body-level ErrorType) | clean (HasErrors gate) | clean (HasErrors gate) |

After the fix, B2 produces (truncated):

```
--- B2 with one unresolved parameter type ---
Emit success: False
Error count: 4
  error CS0246: The type or namespace name 'NotDefinedType' could not be found
  ...
```

The user-facing behaviour goes from "compiler process dies with `NotSupportedException`" to "compile fails with a `CS0246` pointing at the offending source line" — which is exactly what the issue's Suggested-Fix-Scope §2 asks for.

### 6. Phased plan (Phase 1 = this plan; Phase 2+ = implementation gated on approval)

| Phase | Deliverable | Notes |
|---|---|---|
| 1 | This plan + reviewer-approved approach | You are reading it |
| 2 | Branch from `origin/master` (inherits WI-93). Apply the `ErrorType` arm above. | One file change: `SymbolSerializer.cs` |
| 3 | Regression test in `Test/Compiler/NScript.Csc.Lib.Test/` that drives the B2 scenario via `SerializationHelper.InjectIntoCompilation` and asserts `Emit().Success == false` with at least one `Diagnostic.Severity == Error` — **no `NotSupportedException`** | New test file. Note: `NScript.Csc.Lib.Test.csproj` is missing `Microsoft.NET.Test.Sdk` (commented out at line 31) → tests in this project don't currently run. WI-93's findings flagged the same. Will re-enable in this PR scope. |
| 4 | Manual smoke against AgentWorkBench's `WebSocketLogSink.cs` repro (developer machine, since Windows-only V8 framework build) | Confirms the production crash path |
| 5 | PR opened against `master` referencing this comment; expect single-iteration review | No proto/serializer surface changes; no AST-shape changes |

### 7. Risks & non-goals

- **Risk: silent broken DLL produced when only an embedded-AST consumer is downstream.** Mitigation: `Emit().Success` is `false` whenever any error diagnostic is present (the placeholder doesn't change that), so the DLL is not written by the existing flow at `SerializationHelper.cs:53-58`. Confirmed empirically — `outputStream` is closed but the existing code does not commit the placeholder-bearing AST resource on `Success == false`.
- **Risk: false-positive — non-error shape with `SymbolKind.ErrorType` for some legitimate Roslyn future surface.** Considered low: `SymbolKind.ErrorType` is documented as "unresolved/error" by Roslyn and is unconditionally a binding failure today.
- **Non-goal: fixing the upstream binding problem itself** (e.g., why a user's project has an unresolved type) — that's the user's responsibility; this WI is about making the compiler fail cleanly when it happens.
- **Non-goal: addressing other unsupported `TypeSymbol` kinds** (`FunctionPointerType`, `DynamicType`). Leave WI-93's enriched throw as a backstop; open a follow-up if/when those surface.

### 8. Files touched (anticipated)

- `Sources/Compiler/NScript.Csc.Lib/SymbolSerializer.cs` — add `ErrorType` arm (~10 lines)
- `Test/Compiler/NScript.Csc.Lib.Test/NScript.Csc.Lib.Test.csproj` — re-enable `Microsoft.NET.Test.Sdk`
- `Test/Compiler/NScript.Csc.Lib.Test/ErrorTypeLeakTests.cs` *(new)* — Roslyn-driven regression for B2-shape signatures

---

> **To approve**, reply with `/approve` on its own line.
> **To approve with adjustments**, reply with `/approve-with-changes: <your instructions>`.
> **To request changes or answer questions**, reply in prose — the daemon will re-plan.

--- a/Sources/Compiler/NScript.Csc.Lib/SymbolSerializer.cs
+++ b/Sources/Compiler/NScript.Csc.Lib/SymbolSerializer.cs
@@ -211,6 +211,24 @@ namespace NScript.Csc.Lib
                 { PointedAtType = GetTypeSpecSer(((PointerTypeSymbol)type).PointedAtType) };
             }
 
+            if (type.Kind == SymbolKind.ErrorType)
+            {
+                // WI-94: An ErrorTypeSymbol reaches here when a method signature
+                // (param/return type) cannot be resolved (e.g. CS0246 unknown
+                // type). The body-level HasErrors gate in
+                // SerializationHelper.InjectIntoCompilation does not catch this
+                // because the owning method's body may bind cleanly. Returning
+                // a placeholder lets Roslyn finish emit; Emit() then reports
+                // Success=false with the underlying diagnostics, instead of
+                // crashing with NotSupportedException deep inside our hook.
+                return new TypeSpecSer
+                {
+                    Name = "<error>",
+                    Namespace = null,
+                    Module = null
+                };
+            }
+
             if (type.ContainingModule == null)
             {
                 return new TypeSpecSer
@@ -236,14 +254,14 @@ namespace NScript.Csc.Lib
 
             if (type.Kind != SymbolKind.NamedType)
             {
-                // WI-93: Surface enough context that the next regression in this
-                // arm is identifiable without an instrumentation cycle. Common
-                // ways to reach here include ErrorTypeSymbol (binding error),
-                // FunctionPointerType, and future C# surface (e.g. ref-struct
-                // shapes Roslyn introduces later). The primary HasErrors gate in
-                // SerializationHelper.InjectIntoCompilation suppresses the
-                // ErrorType case; this throw remains a backstop for genuinely
-                // unsupported shapes.
+                // WI-93/WI-94: Surface enough context that the next regression
+                // in this arm is identifiable without an instrumentation cycle.
+                // ErrorTypeSymbol is handled by the dedicated arm above; the
+                // body-level HasErrors path is filtered in
+                // SerializationHelper.InjectIntoCompilation. This throw remains
+                // a backstop for genuinely unsupported shapes such as
+                // FunctionPointerType or future C# surface (e.g. new ref-struct
+                // shapes Roslyn may introduce).
                 throw new NotSupportedException(
                     string.Format(
                         "NScript.Csc.Lib.SymbolSerializer.Serialize: unsupported TypeSymbol Kind='{0}', RuntimeType='{1}', FQN='{2}', ContainingType='{3}'.",

--- a/Test/Compiler/NScript.Csc.Lib.Test/ErrorTypeLeakTests.cs
+++ b/Test/Compiler/NScript.Csc.Lib.Test/ErrorTypeLeakTests.cs
@@ -1,0 +1,238 @@
+namespace NScript.Csc.Lib.Test
+{
+    using System;
+    using System.IO;
+    using System.Linq;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+    /// <summary>
+    /// Regression coverage for WI-94: an <c>ErrorTypeSymbol</c> reaching
+    /// <c>SymbolSerializer.Serialize(TypeSymbol)</c> via the METHOD SIGNATURE
+    /// path (param/return type), not the method body. WI-93's
+    /// <c>HasErrors</c> gate in <c>SerializationHelper.InjectIntoCompilation</c>
+    /// is body-only, so it does not fire when a sibling method's body binds
+    /// cleanly but its signature references an undefined type. The fix is a
+    /// dedicated <c>ErrorType</c> arm in <c>Serialize(TypeSymbol)</c> that
+    /// returns a placeholder <c>TypeSpecSer</c> instead of throwing
+    /// <c>NotSupportedException</c>.
+    /// </summary>
+    [TestClass]
+    public class ErrorTypeLeakTests
+    {
+        [TestMethod]
+        public void SignatureLevel_UnknownParameterType_DoesNotThrow()
+        {
+            // The method body is empty (no body-level binding errors), but the
+            // parameter type `NotDefinedType` is unresolved. Roslyn binds the
+            // parameter to MissingMetadataTypeSymbol+TopLevel (Kind=ErrorType).
+            // SerializationHelper's HasErrors gate sees bodyHasErrors=False and
+            // does NOT skip the callback; SymbolSerializer.GetMethodSpecId then
+            // serializes the signature and hits the ErrorType arm.
+            const string source = @"
+namespace Repro
+{
+    public class B2
+    {
+        public void DoSomething(NotDefinedType x) { }
+    }
+}";
+            var outcome = Pipeline.Invoke(source, "SigParamRepro");
+
+            Assert.IsNull(
+                outcome.ThrownException,
+                "Signature-level ErrorType must not crash the serializer. WI-94 regression. Thrown: "
+                    + outcome.ThrownException);
+            Assert.IsFalse(
+                outcome.Success,
+                "Emit must report failure when a parameter type fails to bind.");
+            Assert.IsTrue(
+                outcome.Diagnostics.Any(d =>
+                    d.Severity == DiagnosticSeverity.Error
+                    && (d.Id == "CS0246" || d.Id == "CS0234" || d.Id == "CS0117")),
+                "Expected a Roslyn type-resolution diagnostic. Actual: "
+                    + string.Join(", ", outcome.Diagnostics.Select(d => d.Id + ":" + d.GetMessage())));
+        }
+
+        [TestMethod]
+        public void SignatureLevel_UnknownReturnType_DoesNotThrow()
+        {
+            // Mirror of the parameter case for the return-type path. Roslyn
+            // serializes return type before parameters; both flow through the
+            // same Serialize(TypeSymbol) entry.
+            const string source = @"
+namespace Repro
+{
+    public class B2
+    {
+        public NotDefinedType GetSomething() { return default(NotDefinedType); }
+    }
+}";
+            var outcome = Pipeline.Invoke(source, "SigReturnRepro");
+
+            Assert.IsNull(
+                outcome.ThrownException,
+                "Signature-level ErrorType (return type) must not crash. Thrown: "
+                    + outcome.ThrownException);
+            Assert.IsFalse(outcome.Success);
+            Assert.IsTrue(
+                outcome.Diagnostics.Any(d =>
+                    d.Severity == DiagnosticSeverity.Error
+                    && (d.Id == "CS0246" || d.Id == "CS0234")),
+                "Expected CS0246/CS0234 for unresolved return type. Actual: "
+                    + string.Join(", ", outcome.Diagnostics.Select(d => d.Id + ":" + d.GetMessage())));
+        }
+
+        [TestMethod]
+        public void SignatureLevel_MixedWithHealthySibling_CapturesHealthyOnly()
+        {
+            // Pair an unresolved-signature method with a healthy sibling. The
+            // healthy method must still be captured (HasErrors gate is
+            // per-callback) — the broken signature must not poison the rest of
+            // the compilation.
+            const string source = @"
+namespace Repro
+{
+    public class B2
+    {
+        public void Broken(NotDefinedType x) { }
+        public int Healthy(int y) { return y + 1; }
+    }
+}";
+            var outcome = Pipeline.Invoke(source, "MixedSigRepro");
+
+            Assert.IsNull(
+                outcome.ThrownException,
+                "Mixed compilation with sig-level ErrorType must not crash. Thrown: "
+                    + outcome.ThrownException);
+            Assert.IsFalse(outcome.Success);
+            Assert.IsTrue(
+                outcome.MethodCount >= 1,
+                "Healthy sibling method body must still be captured. Captured: "
+                    + outcome.MethodCount);
+        }
+
+        [TestMethod]
+        public void SignatureLevel_GenericArgumentIsErrorType_DoesNotThrow()
+        {
+            // The unresolved type appears as a generic argument inside a
+            // resolved generic (List<T>). NamedType is serialized normally;
+            // its TypeArguments recurse through GetTypeSpecSer → the ErrorType
+            // arm fires for the inner argument.
+            const string source = @"
+namespace Repro
+{
+    public class B2
+    {
+        public void Take(System.Collections.Generic.List<NotDefinedType> xs) { }
+    }
+}";
+            var outcome = Pipeline.Invoke(source, "GenericArgRepro");
+
+            Assert.IsNull(
+                outcome.ThrownException,
+                "Generic argument that is an ErrorType must not crash. Thrown: "
+                    + outcome.ThrownException);
+            Assert.IsFalse(outcome.Success);
+        }
+
+        [TestMethod]
+        public void SignatureLevel_ArrayOfErrorType_DoesNotThrow()
+        {
+            // ArrayTypeSymbol whose ElementType is an ErrorType — the array
+            // arm of Serialize recurses on the element, hitting the ErrorType
+            // arm transitively.
+            const string source = @"
+namespace Repro
+{
+    public class B2
+    {
+        public void Take(NotDefinedType[] xs) { }
+    }
+}";
+            var outcome = Pipeline.Invoke(source, "ArrayErrorRepro");
+
+            Assert.IsNull(
+                outcome.ThrownException,
+                "Array element ErrorType must not crash. Thrown: "
+                    + outcome.ThrownException);
+            Assert.IsFalse(outcome.Success);
+        }
+
+        [TestMethod]
+        public void TitleScenario_DeclareThenAssignInTry_DoesNotThrow()
+        {
+            // The issue title names "declare-then-assign-in-try local". WI-93's
+            // body-level HasErrors gate already covers this body-only shape
+            // when the bound expression carries errors. Keep a regression test
+            // here so the title scenario itself is locked in.
+            const string source = @"
+namespace Repro
+{
+    public class B3
+    {
+        public void Send(string s)
+        {
+            UnknownType local;
+            try
+            {
+                local = new UnknownType();
+            }
+            catch
+            {
+                local = null;
+            }
+        }
+    }
+}";
+            var outcome = Pipeline.Invoke(source, "DeclareAssignTryRepro");
+
+            Assert.IsNull(
+                outcome.ThrownException,
+                "Body-level declare-then-assign-in-try with ErrorType must not crash. Thrown: "
+                    + outcome.ThrownException);
+            Assert.IsFalse(outcome.Success);
+        }
+
+        private static class Pipeline
+        {
+            public sealed class Outcome
+            {
+                public bool Success { get; init; }
+                public System.Collections.Immutable.ImmutableArray<Diagnostic> Diagnostics { get; init; }
+                    = System.Collections.Immutable.ImmutableArray<Diagnostic>.Empty;
+                public int MethodCount { get; init; }
+                public Exception ThrownException { get; init; }
+            }
+
+            public static Outcome Invoke(string source, string assemblyName)
+            {
+                var tree = CSharpSyntaxTree.ParseText(source, path: assemblyName + ".cs");
+                var mscorlib = MetadataReference.CreateFromFile(typeof(object).Assembly.Location);
+                var compilation = CSharpCompilation.Create(
+                    assemblyName,
+                    syntaxTrees: new[] { tree },
+                    references: new[] { mscorlib },
+                    options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+
+                try
+                {
+                    var (resources, rv) = SerializationHelper.InjectIntoCompilation(compilation);
+                    using var output = new MemoryStream();
+                    var result = compilation.Emit(output, manifestResources: resources);
+                    return new Outcome
+                    {
+                        Success = result.Success,
+                        Diagnostics = result.Diagnostics,
+                        MethodCount = rv.Count,
+                    };
+                }
+                catch (Exception ex)
+                {
+                    return new Outcome { ThrownException = ex };
+                }
+            }
+        }
+    }
+}

